MobileSdk/src/main/java/com/spruceid/mobile/sdk/GattClient.kt

```diff
@@ -333,101 +348,300 @@ class GattClient(private var callback: GattClientCallback,
+        }
+    }
+
+    /**
+     * Set notifications for a characteristic.  This process is rather more complex than you'd think it would
+     * be, and isn't complete until onDescriptorWrite() is hit; it triggers an async action.
+     */
+    private fun enableNotification(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic?, name: String) {
+        dprint("Enabling notifications on ${name}")
 
-            if (!gatt.setCharacteristicNotification(characteristicServer2Client, true)) {
-                callback.onError(Error("Error setting notification on Server2Client."))
+        if (characteristic == null) {
+            reportError("Error setting notification on ${name}; is null.")
+            return
+        }
+
+        dprint("-- ind: ${characteristic.getProperties()}")
+
+
+        try {
+            if (!gatt.setCharacteristicNotification(characteristic, true)) {
+                reportError("Error setting notification on ${name}; call failed.")
                 return
             }
 
-            val descriptor: BluetoothGattDescriptor =
-                characteristicServer2Client!!.getDescriptor(clientCharacteristicConfigUuid)
+            val descriptor: BluetoothGattDescriptor = characteristic.getDescriptor(clientCharacteristicConfigUuid)
 
             if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                 val res = gatt.writeDescriptor(descriptor, BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE)
                 if(res != BluetoothStatusCodes.SUCCESS) {
-                    callback.onError(Error("Error writing to Server2Client. Code: $res"))
+                    reportError("Error writing to ${name}. Code: $res")
                     return
                 }
             } else {
                 // Above code addresses the deprecation but requires API 33+
                 @Suppress("deprecation")
                 descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
-
+                
                 @Suppress("deprecation")
                 if (!gatt.writeDescriptor(descriptor)) {
-                    callback.onError(Error("Error writing to Server2Client clientCharacteristicConfig: desc."))
+                    reportError("Error writing to ${name} clientCharacteristicConfig: desc.")
                     return
                 }
             }
+        } catch (e: SecurityException) {
+            reportError("Not authorized to enable notification on ${name}")
+        }
+
+        // An onDescriptorWrite() call will come in for the pair of this characteristic and the client
+        // characteristic config UUID when notification setting is complete.
+    }
+
+    /**
+     * Log stuff to the console.
+     */
+    private fun dprint(text: String) {
+        Log.d("[GattClient]", text)
+    }
+
+    /**
+     * Log stuff both to the callback and to the console.
+     */
+    private fun reportLog(text: String) {
+        dprint(text)
```

Comment:
```suggestion
        Log.info("[GattClient]", text)
```

Reply:
Fixed, though I've used Log.d(...) since Log.info(...) doesn't compile. I could have used `Log.i(...)`, but much of what I used here seemed like debug level rather than info level.